unpack-trees.c

```diff
@@ -665,17 +665,24 @@ static void mark_ce_used_same_name(struct cache_entry *ce,
 	}
 }
 
-static struct cache_entry *next_cache_entry(struct unpack_trees_options *o)
+static struct cache_entry *next_cache_entry(struct unpack_trees_options *o, int *hint)
 {
 	const struct index_state *index = o->src_index;
 	int pos = o->cache_bottom;
```

Comment:
I know `cache_bottom` is critical to this change, but a little unsure what it means?

Reply:
As we scan the index, we are also drilling down the trees, so the cache-tree extension (if available) is giving us some idea of the range of index entries corresponding to a given tree. This can be computed dynamically, too. The `cache_bottom` value is storing the expected lower-end of this range. As we recurse into subtrees, the `cache_bottom` is stored in a stack variable somewhere and then the `o->cache_bottom` value is rewritten as the recursion ends.

In general: the `cache_bottom` thing is necessary to make these index scans work properly, but it is complicated because the index is a flat list of files. It's just one more thing that is added on top to try and make sense of that difference with the trees in the object database.

Reply:
*Disclaimer: I'm still learning a lot of this too, so I might get some of this wrong*

`cache_bottom` is a position within the index as a whole used in `unpack_trees` to keep track of what has already been processed and added to the result, whether by just "unpacking" the index entry (copying it from source index -> result) or by merging it with the given tree(s). 

In non-sparse index cases, `cache_bottom` pretty much lines up with `hint` (and serves the same purpose - skipping all the already-unpacked index entries). However, for reasons I still need to fully understand, the `cache_bottom` is not incremented the same way (it stops at, I _think_ the first sparse directory position in the index). The result is that, without `hint`, `next_cache_entry` will start from the (early-position) `cache_bottom` and loop over entries it has already handled.